### PR TITLE
Feature/multi select cell

### PR DESCRIPTION
**Pull request by shaneporter**
Pull request for multi-select cell #658. Note that crop has been implemented as 'toggle selection -> delete'.

### CODE REVIEW

**Comment by gordonwoodhull** [comment]
This works great and the code is very clear!

I wouldn't have thought of making the selection state part of the notebook model, but that makes a lot of sense.

The main suggestions I have are aesthetic:
 - the selection bar is logically part of the notebook and shouldn't blend into the title bars of the side panels. I appreciate that the current design creates clean lines, but here, combining the selection bar with the side panels confuses me. I think it should be a lighter grey (maybe somewhere between the cell commands and the panel titles).
 - the checkbox control can also be shorter IMO. This is the most crucial real estate so if the "outer button" can just fit the checkbox I think that would be a win.
 - notice the weird effect when the find and replace panel is open (ctrl-F) - I think this is the correct order (selection feels conceptually closer to the cells than find?) but it underscores that the selection bar is part of the notebook, not part of the navbar.
 - the equivalent control in gmail uses the entire outer button to trigger the dropdown. That saves one line of clutter, and is a sensible action for clicking outside the checkbox but inside the button (which currently does nothing).
 - the yellow is a bit violent - maybe we can use something more subtle, like a pale green? and maybe it should just fill the cell command bar and the gutter, not the code and results? I'm not sure. Right now it interferes with editing because it is so loud. I wouldn't want people to have to deselect just to continue editing code.

**Comment by gordonwoodhull** [comment]
I also have a round of UI improvements we can consider after this is merged:

 - I notice that clicking the cell name toggles selection, which is great, but I think clicking the rest of the cell bar should do the same thing. (The cell name and the cell bar are the same object.)
 - I would expect clicking a cell name/cell bar to have a slightly different effect from clicking the checkbox. It should be more like the multiple selection interface: if the cell is already selected, it should have no effect. If it is not selected, then the selection is cleared and only that cell is selected.
 - We'll eventually have to consider if we need [multisortable](http://jsfiddle.net/neochief/KWeMM/) - maybe not for this release, but I think it's what people will expect, and it's also helpful, e.g. to move a bunch of currently dispersed cells to the beginning or end. Multisortable is not part of core jQuery UI, so I don't know if there is a stable implementation.

**Comment by gordonwoodhull** [comment]
Great work!

File: htdocs/js/notebook/notebook_model.js [hunk 1]
```diff
@@ -174,6 +174,35 @@ Notebook.create_model = function()
...
         },
+        remove_selected_cells: function() {
+            var that = this;
+            _.chain(this.cells)
+            .filter(function(cell) {
+                return cell.is_selected();
+            })
+            .each(function(cell) {
+                that.remove_cell(cell);
+            });
```
Comment (gordonwoodhull) [inline htdocs/js/notebook/notebook_model.js]: It doesn't really matter, but underscore isn't needed here - `this.cells.filter(...).forEach(...)` is fine.

Comment (gordonwoodhull) [inline htdocs/js/notebook/notebook_model.js]: (not that our code has any consistent style about this sort of thing either)

**Comment by shaneporter** [comment]
image

This shows an updated selection bar (d222aa8, 69ced7f and ac2833e) and selected cell change (not committed at time of writing.) Not sure what I think about it so welcome a second opinion. In fact, I've just thought of an alternative, which I will add in another comment.

I haven't changed the colour, preferring not to change to green since this might interfere with the cell status (circled) which is also green.

I wasn't 100% about what was necessary for the selection bar, but have made it lighter. Visually, it is separate from the side panels' title bars. The checkbox selection area is now slimmer.

**Comment by gordonwoodhull** [comment]
Yeah, we can continue to revise it, but I would consider each of those points fixed with the design you show above.

**Comment by gordonwoodhull** [comment]
With the selection bar in that size and color, it enhances the illusion that the notebook is "above" the left and right columns. I like it.

**Comment by shaneporter** [comment]
Hmm... that yellowish box shadow didn't really work across a range of cell sizes. For larger ones it was ok, but for smaller ones, it was rather overpowering.

So, I've adopted your suggestion of filling the command bar and gutter:

image

It's fairly unobtrusive, but obvious at the same time. (Committed with 8fa665a).

**Comment by shaneporter** [comment]
One of your post-merge considerations, 'short cut for delete' had been implemented. 

Invert and crop/cull shortcuts shouldn't take too long to implement. Can you suggest their keyboard shortcuts?

**Comment by gordonwoodhull** [comment]
That's a bigger question than it sounds - I think we need to look at shortcuts from a global level as part of #300.

As a placeholder, let's use what Preview uses on the Mac, ctrl/cmd-shift-I for invert, and ctrl/cmd-K for crop.

**Comment by shaneporter** [comment]
Sounds good to me.

**Comment by gordonwoodhull** [comment]
I'm not sure if I agree with myself about the bar and gutter (off balance?) but it's a good start. As you said, it's unobtrusive and obvious, so it's ready for people to test. 

Let's merge this tomorrow and then continue to iterate once we get some feedback from users.

**Comment by gordonwoodhull** [comment]
Oops, I missed that final commit. Merging now.

BTW, we switched to grunt instead of make (nice!): please rebase any feature branches you're working on (or merge from develop).

**Comment by gordonwoodhull** [comment]
Merged!

There are a lot of implications we'll have to work through, but I can tell this is going to be very, very helpful for folks. Thank you @shaneporter!

Eventually I'll open separate issues for some of the stuff discussed above. Let's get a little feedback first.

**Pull request closed without merging**